torch/nn/modules/module.py

```diff
@@ -534,12 +560,26 @@ def load_state_dict(self, state_dict, strict=True):
                                        'whose dimensions in the checkpoint are {}.'
                                        .format(name, own_state[name].size(), param.size()))
             elif strict:
-                raise KeyError('unexpected key "{}" in state_dict'
-                               .format(name))
+                unexpected.append(name)
         if strict:
             missing = set(own_state.keys()) - set(state_dict.keys())
+            # pass the mismatch info to submodules so that they have a chance to
+            # raise a custom class-specific error
+            for name in unexpected:
+                submodule_key_mismatch(name, False)
+            for name in missing:
+                submodule_key_mismatch(name, True)
+            error = ''
+            if len(unexpected) > 0:
+                error += ('Unexpected key(s) in state_dict: {}. '.format(
+                    ', '.join('"{}"'.format(k) for k in unexpected)))
             if len(missing) > 0:
-                raise KeyError('missing keys in state_dict: "{}"'.format(missing))
+                error += ('Missing key(s) in state_dict: {}. '.format(
+                    ', '.join('"{}"'.format(k) for k in unexpected)))
+            if len(error) > 0:
+                raise KeyError(error + 'If this is expected, set strict=False '
```

Comment:
Srsly, we should avoid recommending `strict=False`. It's better to fix the dicts explicitly